pkg/extensions/unpack.go

```diff
@@ -0,0 +1,295 @@
+	dirExists, err := afero.DirExists(fs, resourcesDir)
+	if !dirExists || err != nil {
+		// the resources dir doesn't appear to exist, this isn't an error but return nothing
+		return v1alpha1.NewCRDList(), "", nil
+	}
+
+	// walk the resources dir and find all the CRD files
+	resourcesFiles, err := findResourcesFiles(fs, resourcesDir)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// now that we have found all the resource files, process each one in the list
+	var sb strings.Builder
+	crdList = v1alpha1.NewCRDList()
+	for _, rf := range resourcesFiles {
+		if err := readResourceFile(fs, rf, crdList, &sb); err != nil {
+			return nil, "", err
+		}
+	}
+
+	return crdList, sb.String(), nil
+}
+
+func findResourcesFiles(fs afero.Fs, resourcesDir string) ([]string, error) {
+	resourcesFiles := []string{}
+	err := afero.Walk(fs, resourcesDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			// just proceed to next entry on walk
+			return nil
+		}
+		if matched, err := filepath.Match(resourceFileNamePattern, filepath.Base(path)); matched && err == nil {
+			// current item on walk matches the resource file name pattern, include it in the list of matches
+			resourcesFiles = append(resourcesFiles, path)
+		}
+		return nil
+	})
+	if err != nil {
+		// walking for resources encountered an error that halted it, surface this error
+		return nil, err
+	}
+
+	return resourcesFiles, nil
+}
+
+func readResourceFile(fs afero.Fs, rf string, crdList *v1alpha1.CRDList, sw io.StringWriter) error {
+	b, err := afero.ReadFile(fs, rf)
+	if err != nil {
+		// we weren't able to read the current resource file, surface this error
+		return err
+	}
+
+	// unmarshal the raw resource file content into a CRD type
+	var crd apiextensions.CustomResourceDefinition
+	if err := yaml.Unmarshal(b, &crd); err != nil {
+		return err
+	}
+
+	// add the CRD type meta to the list
+	// TODO(jbw976): handle cases where the CRD has multiple versions associated with it
+	crdGVK := schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: crd.Spec.Version,
+		Kind:    crd.Spec.Names.Kind,
+	}
+	crdTypeMeta := metav1.TypeMeta{
+		Kind:       crdGVK.Kind,
+		APIVersion: crdGVK.GroupVersion().String(),
+	}
+	crdList.Owned = append(crdList.Owned, crdTypeMeta)
+
+	// add the raw resource file content to the string builder
+	if _, err := sw.WriteString(yamlSeparator + string(b)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readIcons(fs afero.Fs, root string) ([]v1alpha1.IconSpec, error) {
+	// look for icon files that start with the standard icon file name pattern
+	matches, err := afero.Glob(fs, filepath.Join(root, iconFileNamePattern))
+	if err != nil {
+		return nil, err
+	}
+
+	icons := make([]v1alpha1.IconSpec, len(matches))
+	for i, m := range matches {
+		// open the icon file for reading from
+		f, err := fs.Open(m)
+		if err != nil {
+			return nil, err
+		}
+		defer func() { _ = f.Close() }()
```

Comment:
Defer inside the loop. Check this out https://blog.learngoprogramming.com/gotchas-of-defer-in-go-1-8d070894cb01 for the problem and possible solutions.

Reply:
good catch! i've followed the guidance to wrap the loop body in an anonymous func so the defers will execute per iteration of the loop